src/gpuarray_reduction.c

```diff
@@ -0,0 +1,804 @@
+	int         ret;
+	size_t      blockSize[3]   = {1,1,1};
+	size_t      gridSize [3]   = {1,1,1};
+	size_t      chunkSize[3]   = {1,1,1};
+	gpudata*    srcStepsGD     = 0, *srcSizeGD        = 0, *chunkSizeGD = 0,
+	            *dstMaxStepsGD = 0, *dstArgmaxStepsGD = 0;
+	gpucontext* ctx            = GpuArray_context(src);
+	
+	
+	/**
+	 * Schedule the kernel.
+	 * 
+	 * This implies choosing the block, grid and chunk size appropriately.
+	 */
+	
+	scheduleMaxAndArgmax(k, src, reduxLen, reduxList,
+	                     blockSize, gridSize, chunkSize);
+	
+	
+	/**
+	 * Argument Marshalling. This the grossest gross thing in here.
+	 */
+	
+	srcStepsGD       = gpudata_alloc(ctx, src->nd * sizeof(size_t),
+	                                 src->strides,
+	                                 GA_BUFFER_READ_ONLY|GA_BUFFER_INIT, &ret);
+	if(ret){goto releaseGpudata;}
+	srcSizeGD        = gpudata_alloc(ctx, src->nd * sizeof(size_t),
+	                                 src->dimensions,
+	                                 GA_BUFFER_READ_ONLY|GA_BUFFER_INIT, &ret);
+	if(ret){goto releaseGpudata;}
+	chunkSizeGD      = gpudata_alloc(ctx, src->nd * sizeof(size_t),
+	                                 chunkSize,
+	                                 GA_BUFFER_READ_ONLY|GA_BUFFER_INIT, &ret);
+	if(ret){goto releaseGpudata;}
+	dstMaxStepsGD    = gpudata_alloc(ctx, (src->nd - reduxLen) * sizeof(size_t),
+	                                 dstMax->strides,
+	                                 GA_BUFFER_READ_ONLY|GA_BUFFER_INIT, &ret);
+	if(ret){goto releaseGpudata;}
+	dstArgmaxStepsGD = gpudata_alloc(ctx, (src->nd - reduxLen) * sizeof(size_t),
+	                                 dstArgmax->strides,
+	                                 GA_BUFFER_READ_ONLY|GA_BUFFER_INIT, &ret);
+	if(ret){goto releaseGpudata;}
+	
+	
+	struct MaxAndArgmaxArgs{
+		gpudata*  src;
+		size_t    srcOff;
+		gpudata*  srcSteps;
+		gpudata*  srcSize;
+		gpudata*  chunkSize;
+		gpudata*  dstMax;
+		size_t    dstMaxOff;
+		gpudata*  dstMaxSteps;
+		gpudata*  dstArgmax;
+		size_t    dstArgmaxOff;
+		gpudata*  dstArgmaxSteps;
+	} argstr = {
+		src->data,
+		src->offset,
+		srcStepsGD,
+		srcSizeGD,
+		chunkSizeGD,
+		dstMax->data,
+		dstMax->offset,
+		dstMaxStepsGD,
+		dstArgmax->data,
+		dstArgmax->offset,
+		dstArgmaxStepsGD
+	};
+	
+	void* args[] = {
+		(void*)&argstr.src,
+		(void*)&argstr.srcOff,
+		(void*)&argstr.srcSteps,
+		(void*)&argstr.srcSize,
+		(void*)&argstr.chunkSize,
+		(void*)&argstr.dstMax,
+		(void*)&argstr.dstMaxOff,
+		(void*)&argstr.dstMaxSteps,
+		(void*)&argstr.dstArgmax,
+		(void*)&argstr.dstArgmaxOff,
+		(void*)&argstr.dstArgmaxSteps
+	};
+	
+	
+	/**
+	 * Call kernel, release arguments and return error code
+	 */
+	
+	ret = GpuKernel_call(k, 3, blockSize, gridSize, 0, args);
+	releaseGpudata:
+	gpudata_release(srcStepsGD);
+	gpudata_release(srcSizeGD);
+	gpudata_release(chunkSizeGD);
+	gpudata_release(dstMaxStepsGD);
+	gpudata_release(dstArgmaxStepsGD);
```

Comment:
@abergeron To note what has been said in email, this is apparently not safe, since `gpudata_release()` chokes on `NULL`. IMHO it would be worthwhile to add a quick `if(ptr){...}` to `gpudata_release()`, to imitate `free(NULL)`'s safe semantics and to simplify code that uses several `gpudata_alloc()` calls.